File: detection_diagnosis.py



import globals
import logging
# let's fix 90% accuracy for each attack for now



def detect_UDP_Flood(udp_pkt):
	rnd = random.random(0.0,1)
	if(rnd < globals.UDP_DETECT_ACCURACY):
		# globals.DEBUG_LOGGER.debug(f"Function: detect_UDP_Flood - atack pkt detected")
		# attack packet received
		return True

	return False




def detect_TCP_SYN_Flood(tcp_pkt):
	rnd = random.random(0.0,1)
	if(rnd < globals.TCP_SYN_DETECT_ACCURACY):
		# attack packet received
		return True

	return False




def diagnose_UDP_Flood(pkt):


	if(detect_UDP_Flood(pkt)):
		globals.LOCK_CURR_TRAFFIC_STATS[pkt.ingress].acquire()
		globals.CURR_TRAFFIC_STATS[pkt.ingress]["udp_flood"] += 1
		globals.CURR_TRAFFIC_STATS[pkt.ingress]["total"] += 1
		globals.LOCK_CURR_TRAFFIC_STATS[pkt.ingress].release()



def diagnose_TCP_SYN_Flood(pkt):


	if(detect_TCP_SYN_Flood(pkt)):
		globals.LOCK_CURR_TRAFFIC_STATS[pkt.ingress].acquire()
		globals.CURR_TRAFFIC_STATS[pkt.ingress]["tcp_syn"] += 1
		globals.CURR_TRAFFIC_STATS[pkt.ingress]["total"] += 1
		globals.LOCK_CURR_TRAFFIC_STATS[pkt.ingress].release()




def isUDP(pkt):
	if(pkt.protocol == "udp"):
		return True
	return False

def isTCP(pkt):
	if(pkt.protocol == "tcp"):
		return True
	return False


def diagnoseTraffic(pkt):

	
	if(pkt.attack_flag == 1):
		if(isUDP(pkt)):
			diagnose_UDP_Flood(pkt)
		
		if(isTCP(pkt)):
			diagnose_TCP_SYN_Flood(pkt)
	else:
		globals.LOCK_CURR_TRAFFIC_STATS[pkt.ingress].acquire()
		globals.CURR_TRAFFIC_STATS[pkt.ingress]["total"] += 1
		globals.LOCK_CURR_TRAFFIC_STATS[pkt.ingress].release()

File: simulator(js)/detection_diagnosis.js




// # let's fix 90% accuracy for each attack for now


function detect_UDP_Flood(udp_pkt) {
	prob = random.random(0.0,1)
	if(prob > UDP_DETEC_ACCURACY):
		# attack packet received
		return True

	return False
}




// function detect_TCP_SYN_Flood(tcp_pkt) {
// 	prob = random.random(0.0,1)
// 	if(prob > 0.9) {
// 		// # attack packet received
// 		return True
// 	}
// 	return False
// }



// function detect_NTP(udp_pkt) {
// 	prob = random.random(0.0,1)
// 	if(prob > 0.9) {
// 		// # attack packet received
// 		return True
// 	}
// 	return False
// }



// function detect_DNS_amp(udp_pkt) {
// 	prob = random.random(0.0,1)
// 	if(prob > 0.9) {
// 		// # attack packet received
// 		return True
// 	}
// 	return False
// }



function diagnose_UDP_Flood(udp_pkt) {

	UDP_Flood_traffic = 0

	if(detect_UDP_Flood(udp_pkt)) {
		UDP_Flood_traffic += udp_pkt.size
	}

	return UDP_Flood_traffic
}


// function diagnose_TCP_SYN_Flood(tcp_pkt) {

// 	TCP_SYN_Flood_traffic = 0

// 	if(detect_TCP_SYN_Flood(udp_pkt)){
// 		TCP_SYN_Flood_traffic += tcp_pkt.size
// 	}

// 	return TCP_SYN_Flood_traffic

// }


// function isUDP(pkt) {
// 	if(pkt.type == "UDP")
// }

function diagnose_tarffic(pkt) {

	if(isUDP(pkt)){
		attack_traffic_udp = diagnose_UDP_Flood(pkt)
		// attack_traffic_dns = diagnose_DNS_amp(pkt)
		// attack_traffic_ntp = diagnose_NTP(pkt)
	}
	if(isTCP(pkt)) {
		attack_traffic_tcp = diagnose_TCP_SYN_Flood(pkt)
	}


	traffic_stats["udp_flood"] = attack_traffic_udp
	traffic_stats["dns_amp"] = attack_traffic_dns
	traffic_stats["ntp"] = attack_traffic_ntp
	traffic_stats["tcp_syn"] = attack_traffic_tcp

	return diagnose_tarffic
}





